File: MergeSort/mergesort.c
#include <stdio.h>
void merge(int a[], int left, int mid, int right)
{

    int n1 = mid - left + 1;
    int n2 = right - mid;

    int L[n1], M[n2];

    for (int i = 0; i < n1; i++)
        L[i] = a[left + i];
    for (int j = 0; j < n2; j++)
        M[j] = a[mid + 1 + j];

    int i, j, k;
    i = 0;
    j = 0;
    k = left;

    while (i < n1 && j < n2)
    {
        if (L[i] <= M[j])
        {
            a[k] = L[i];
            i++;
        }
        else
        {
            a[k] = M[j];
            j++;
        }
        k++;
    }

    while (i < n1)
    {
        a[k] = L[i];
        i++;
        k++;
    }

    while (j < n2)
    {
        a[k] = M[j];
        j++;
        k++;
    }
}

void mergeSort(int a[], int l, int right)
{
    if (l < right)
    {

        int m = l + (right - l) / 2;

        mergeSort(a, l, m);
        mergeSort(a, m + 1, right);

        merge(a, l, m, right);
    }
}

void printArray(int a[], int e)
{
    for (int i = 0; i < e; i++)
        printf("%d ", a[i]);
    printf("\n");
}

int main()
{
    int e;
    int a[e];
    printf("Enter the no of array: ");
    scanf("%d", &e);
    printf("Enter the elements: ");
    for(int i = 0; i < e; i++){
        scanf("%d", &a[i]);
    }
    mergeSort(a, 0, e - 1);

    printf("Sorted array: \n");
    printArray(a, e);
}

File: MergeSort/mergesort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)

    return merge(left_half, right_half)

def merge(left, right):
    result = []
    left_idx, right_idx = 0, 0

    while left_idx < len(left) and right_idx < len(right):
        if left[left_idx] < right[right_idx]:
            result.append(left[left_idx])
            left_idx += 1
        else:
            result.append(right[right_idx])
            right_idx += 1

    result.extend(left[left_idx:])
    result.extend(right[right_idx:])
    
    return result

# Input from the user
try:
    input_str = input("Enter a list of numbers separated by spaces: ")
    input_list = list(map(int, input_str.split()))
except ValueError:
    print("Invalid input. Please enter a list of numbers.")
    exit()

# Sort the user input using merge sort
sorted_list = merge_sort(input_list)

# Display the sorted list
print("Sorted list:", sorted_list)
